sw/device/lib/otbn.h

```diff
@@ -0,0 +1,248 @@
+  /**
+   * End of OTBN instruction memory.
+   */
+  const char *imem_end;
+  /**
+   * Start of OTBN data memory.
+   */
+  const char *dmem_start;
+  /**
+   * End of OTBN data memory.
+   */
+  const char *dmem_end;
+} otbn_app_t;
+
+/**
+ * A pointer to a symbol in OTBN's instruction or data memory.
+ *
+ * Use `OTBN_USE_PTR()` together with `OTBN_PTR_T_INIT()` to initialize this
+ * structure.
+ */
+typedef struct otbn_ptr {
+  /**
+   * Address of the symbol in normal CPU address space.
+   */
+  const char *ptr_in_cpu_addr_space;
+} otbn_ptr_t;
+
+/**
+ * The result of an OTBN operation.
+ */
+typedef enum otbn_result {
+  /**
+   * The operation succeeded.
+   */
+  kOtbnOk = 0,
+  /**
+   * An unspecified failure occurred.
+   */
+  kOtbnError = 1,
+  /**
+   * The precondition check for an argument passed into the function failed.
+   */
+  kOtbnBadArg = 2,
+  /**
+   * The chosen application is not part of the loaded application.
+   */
+  kOtbnFuncNotInApp = 3,
+  /**
+   * No application is loaded.
+   */
+  kOtbnAppNotLoaded = 4,
+  /**
+   * The execution of the application on OTBN failed.
+   *
+   * More specific error information can be obtained with
+   * `dif_otbn_get_err_code()`.
+   */
+  kOtbnExecutionFailed = 5,
+} otbn_result_t;
+
+/**
+ * OTBN context structure.
+ *
+ * Use `otbn_init()` to initialize.
+ */
+typedef struct otbn {
+  /**
+   * The OTBN DIF to access the hardware.
+   */
+  dif_otbn_t dif;
+
+  /**
+   * The application loaded or to be loaded into OTBN.
+   */
+  const otbn_app_t *app;
+
+  /**
+   * Is the application loaded into OTBN?
+   */
+  bool app_is_loaded;
+} otbn_t;
+
+/**
+ * Make an embedded OTBN application image available for use.
+ *
+ * Make symbols available that indicate the start and the end of instruction
+ * and data memory regions, as they are stored in the device memory.
+ *
+ * @param app_name name of the application to load, which is typically the
+ *                 name of the main (assembly) source file.
+ */
+#define OTBN_USE_APP(app_name)                            \
+  extern const char _otbn_app_##app_name##__imem_start[]; \
+  extern const char _otbn_app_##app_name##__imem_end[];   \
+  extern const char _otbn_app_##app_name##__dmem_start[]; \
+  extern const char _otbn_app_##app_name##__dmem_end[];
+
+/**
+ * Initialize the OTBN application information structure
+ *
+ * After making all required symbols from the application image available
+ * through `OTBN_USE_APP()`, use this macro to initialize an `otbn_app_t`
+ * struct with those symbols.
+ *
+ * @param app_name name of the application to load
+ * @see OTBN_USE_APP()
+ */
+#define OTBN_APP_T_INIT(app_name)                     \
+  {                                                   \
+    .imem_start = _otbn_app_##app_name##__imem_start, \
+    .imem_end = _otbn_app_##app_name##__imem_end,     \
+    .dmem_start = _otbn_app_##app_name##__dmem_start, \
+    .dmem_end = _otbn_app_##app_name##__dmem_end,     \
+  }
```

Comment:
Do you think it makes sense to uinify these macros into one, e.g.
```c
OTBN_APP(my_cool_app, kMyCoolApp);
```
which would generate both the external constants and a static const with the app struct in it?

Reply:
I did try that out initially, but went the other direction:
* I don't like to hide the the assignment.
* I like the flexibility to make the variable static or not depending on the use case.